### PR TITLE
Rework SPIRV_LLVM_Translator using augmented platforms.

### PR DESCRIPTION
**Pull request by maleadt**
Makes it possible to install a single version of the package, automatically selecting the correct artifact in accordance with the available LLVM version.

Depends on https://github.com/JuliaPackaging/BinaryBuilder.jl/pull/1128

### CODE REVIEW

**Comment by maleadt** [comment]
Seems to work beautifully:

```
$ julia-nightly
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.8.0-DEV.1351 (2022-01-20)
 _/ |\__'_|_|_|\__'_|  |  Commit 05d3e84821 (0 days old master)
|__/                   |

(@v1.8) pkg> activate --temp
  Activating new project at `/tmp/jl_n6VNR1`

(jl_n6VNR1) pkg> add https://github.com/maleadt/SPIRV_LLVM_Translator_unified_jll.jl
    Updating git-repo `https://github.com/maleadt/SPIRV_LLVM_Translator_unified_jll.jl`
    Updating registry at `~/Julia/depot/registries/General`
    Updating git-repo `git@github.com:JuliaRegistries/General.git`
   Resolving package versions...
    Updating `/tmp/jl_n6VNR1/Project.toml`
  [85f0d8ed] + SPIRV_LLVM_Translator_unified_jll v0.1.0+0 `https://github.com/maleadt/SPIRV_LLVM_Translator_unified_jll.jl#main`

julia> using SPIRV_LLVM_Translator_unified_jll

julia> SPIRV_LLVM_Translator_unified_jll.libLLVMSPIRV
"/home/tim/Julia/depot/artifacts/7ba45e51e5482d6d5162b9962da3294d546e0e8a/lib/libLLVMSPIRVLib.so"

shell> ldd /home/tim/Julia/depot/artifacts/7ba45e51e5482d6d5162b9962da3294d546e0e8a/lib/libLLVMSPIRVLib.so
        libLLVM-13jl.so => not found
```

```
$ julia-1.7
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.7.1 (2021-12-22)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

(@v1.7) pkg> activate --temp
  Activating new project at `/tmp/jl_CbV8u4`

(jl_CbV8u4) pkg> add https://github.com/maleadt/SPIRV_LLVM_Translator_unified_jll.jl
    Updating git-repo `https://github.com/maleadt/SPIRV_LLVM_Translator_unified_jll.jl`
    Updating registry at `~/Julia/depot/registries/General`
    Updating git-repo `git@github.com:JuliaRegistries/General.git`
   Resolving package versions...
    Updating `/tmp/jl_CbV8u4/Project.toml`
  [85f0d8ed] + SPIRV_LLVM_Translator_unified_jll v0.1.0+0 `https://github.com/maleadt/SPIRV_LLVM_Translator_unified_jll.jl#main`

julia> using SPIRV_LLVM_Translator_unified_jll

julia> SPIRV_LLVM_Translator_unified_jll.libLLVMSPIRV
"/home/tim/Julia/depot/artifacts/aa866f6893e9b4f6fb3c4fafad2753aa7c7fbd7c/lib/libLLVMSPIRVLib.so"

shell> ldd /home/tim/Julia/depot/artifacts/aa866f6893e9b4f6fb3c4fafad2753aa7c7fbd7c/lib/libLLVMSPIRVLib.so
        libLLVM-12jl.so => not found
```

**Comment by giordano** [comment]
Mostly out of curiosity, what about loading time of the package?  I expect the platform augmentation stuff to slightly negatively affect it

**Comment by maleadt** [comment]
```
Benchmark 1: julia-1.7 --project -e 'using SPIRV_LLVM_Translator_jll'
  Time (mean ± σ):     103.9 ms ±   6.7 ms    [User: 104.9 ms, System: 123.4 ms]
  Range (min … max):    98.8 ms … 118.9 ms    29 runs
 
Benchmark 1: julia-1.7 --project -e 'using SPIRV_LLVM_Translator_unified_jll'
  Time (mean ± σ):     141.0 ms ±   4.8 ms    [User: 208.4 ms, System: 327.7 ms]
  Range (min … max):   136.5 ms … 154.4 ms    21 runs
 ```

Pretty significant.

**Comment by maleadt** [comment]
I changed the way `build_tarballs` is invoked so that we can leave out the `--register` args, as is done with other packages, but is that still required?

I did so to try and fix local deployment, but it still fails:

```
$ jl --project build_tarballs.jl --debug --verbose x86_64-linux-gnu-cxx11-llvm+12,x86_64-linux-gnu-cxx11-llvm+13 --deploy=maleadt/SPIRV_LLVM_Translator_unified_jll.jl

...

[ Info: Deploying binaries to release SPIRV_LLVM_Translator_unified-v0.1.0+0 on maleadt/SPIRV_LLVM_Translator_unified_jll.jl via `ghr`...
==> Create a new release
--> Uploading: SPIRV_LLVM_Translator_unified.v0.1.0.x86_64-linux-gnu-cxx11-llvm+12.tar.gz

but the during the second build

[ Info: Deploying binaries to release SPIRV_LLVM_Translator_unified-v0.1.0+0 on maleadt/SPIRV_LLVM_Translator_unified_jll.jl via `ghr`...
WARNING: found release (SPIRV_LLVM_Translator_unified-v0.1.0+0). Use existing one.
--> Uploading: SPIRV_LLVM_Translator_unified.v0.1.0.x86_64-linux-gnu-cxx11-llvm+13.tar.gz
--> Uploading: SPIRV_LLVM_Translator_unified.v0.1.0.x86_64-linux-gnu-cxx11-llvm+12.tar.gz
Failed to upload one of assets: one of the goroutines failed: failed to upload asset: /home/tim/Julia/tools/Yggdrasil/S/SPIRV_LLVM_Translator_unified/products/SPIRV_LLVM_Translator_unified.v0.1.0.x86_64-linux-gnu-cxx11-llvm+12.tar.gz: failed to upload release asset: /home/tim/Julia/tools/Yggdrasil/S/SPIRV_LLVM_Translator_unified/products/SPIRV_LLVM_Translator_unified.v0.1.0.x86_64-linux-gnu-cxx11-llvm+12.tar.gz: POST https://uploads.github.com/repos/maleadt/SPIRV_LLVM_Translator_unified_jll.jl/releases/57566045/assets?name=SPIRV_LLVM_Translator_unified.v0.1.0.x86_64-linux-gnu-cxx11-llvm%2B12.tar.gz: 422 Validation Failed [{Resource:ReleaseAsset Field:name Code:already_exists Message:}]
```

Should I similarly leave out the `--deploy` args? Feels like a bit of a hack.

**Comment by giordano** [comment]
You may still need to use [`should_build_platform`](https://github.com/JuliaPackaging/Yggdrasil/blob/2b47d099f1256d7d071e25471187c38f12a060d8/fancy_toys.jl#L6-L35)

**Comment by maleadt** [comment]
> You may still need to use [`should_build_platform`](https://github.com/JuliaPackaging/Yggdrasil/blob/2b47d099f1256d7d071e25471187c38f12a060d8/fancy_toys.jl#L6-L35)

I do so, in populating the `builds` array. That should be sufficient, right?

**Comment by maleadt** [comment]
Added support for detecting LLVM assertions builds (we probably need an API call for this). This looking really nice, I think!

**Comment by maleadt** [comment]
Only passing `--deploy` to the last invocation of `build_tarballs` doesn't seem to work, as then only the hashes for the last upload are correct. Any help here?

File: S/SPIRV_LLVM_Translator_unified/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+"""
+
+# The products that we will ensure are always built
+products = Product[
+    LibraryProduct(["libLLVMSPIRVLib", "LLVMSPIRVLib"], :libLLVMSPIRV, dont_dlopen = true),
+    ExecutableProduct("llvm-spirv", :llvm_spirv),
+]
+
+augment_platform_block = """
+    using Base.BinaryPlatforms
```
Comment (giordano) [inline S/SPIRV_LLVM_Translator_unified/build_tarballs.jl:58]: Why is this needed?  Unless I'm missing something, in this snippet I don't see functions/variables/types from there

Comment (maleadt) [reply S/SPIRV_LLVM_Translator_unified/build_tarballs.jl:58]: Ah yes, good catch, it was using `BinaryPlatforms.add_tag!` before.

**Review by vchuravy — APPROVED** [review]

File: S/SPIRV_LLVM_Translator_unified/build_tarballs.jl (hunk 1 above)
Comment (maleadt) [reply S/SPIRV_LLVM_Translator_unified/build_tarballs.jl:58]: Added it back for a `::Platform` type assertion (to follow the template from the Pkg syntax)
